### ISSUE TITLE
Change  debug build flags from "" (null to "-O0 -g" (upper o+ zero)

### ISSUE THREAD

**Issue by LifWirser** [vegastrike/Vega-Strike-Engine-Source#731]
per
[](url)https://web.mit.edu/gnu/doc/html/gdb_6.html
 [](url)https://web.eecs.umich.edu/~sugih/pointers/summary.html

Two flags must be set
1> turn optimization off via either "-O" or "-O0"
 many sites report using optimising can/does caues unpredictabl behavior in debuggers such as gdb /valgrind
2> set -g
  To request debugging information

Also build type relwithdebinfo should be modified

**Comment by stephengtuggy** [comment]
FWIW, the current code uses `-Og` ("optimize for debugging") and `-g3` ("debug information level 3") for the Debug configuration. That's set on [line 336 in CMakeLists.txt](https://github.com/vegastrike/Vega-Strike-Engine-Source/blob/master/engine/CMakeLists.txt#L336), using [CMake generator expressions](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html) rather than the separate variables per config that were used before.

I will acknowledge that this code is less readable than it could be. In particular, that line is way too long. It should probably be wrapped to fit within 120 columns or something, at a minimum.

**Comment by BenjamenMeyer** [comment]
per https://github.com/vegastrike/Vega-Strike-Engine-Source/issues/731#issuecomment-1227661407 - yeah, we probably should be doing wrapping as:
```
ADD_COMPILE_OPTIONS(
    "$<$<OR:$<CXX_COMPILER_ID:GNU>,$<CXX_COMPILER_ID:Clang>,$<CXX_COMPILER_ID:AppleClang>>:-pipe>"
    "$<$<OR:$<CXX_COMPILER_ID:GNU>,$<CXX_COMPILER_ID:Clang>,$<CXX_COMPILER_ID:AppleClang>>:-Wall>"
    ...
)
```
to make the various ones easier to read.

**Comment by stephengtuggy** [comment]
@LifWirser I've come to the conclusion that you were right about `-O0` being better for debugging than `-Og`. Will plan on incorporating that change soon.

**Comment by BenjamenMeyer** [comment]
We have options for both `-Og` and `-g3` specified in master's CMakeFile.txt for the engine. Is that sufficient to close this?

**Comment by stephengtuggy** [comment]
> We have options for both `-Og` and `-g3` specified in master's CMakeFile.txt for the engine. Is that sufficient to close this?

@LifWirser what's your opinion?

**Comment by LifWirser** [comment]
As events of late, have given me some knowledge:
 basic issue is there is no "firm" standard as to how any compiler/debugger/profiler will functoinn together

 Originally I was looking for ease of compiling /debugging. Since then I gained insight how varying 
 optimization methods require more prosessing to correctrly connect debug sysbols to relevent code.
 It comes to be simular to using "-jx" make flag based on number of cores, add what you system cam handle

  Of course my experienc remind me need for adquite hw resources
 Also how and what order files are linked can add significant resource need just to link ( where all this  
 comes together

**Comment by vincele** [comment]
What I see in `engine/CMakeLists.txt` looks pretty sensible.
I think that the generated debug infos don't have any runtime performance cost, if you're not running under a debugger.
They only take some time to generate when building, and disk space.

**Comment by stephengtuggy** [comment]
> As events of late, have given me some knowledge: basic issue is there is no "firm" standard as to how any compiler/debugger/profiler will functoinn together

Agreed. There is no one agreed-upon standard for the "ideal" compiler debugging settings. Based on that, @LifWirser can this ticket be closed?

**Comment by stephengtuggy** [comment]
If one is concerned about execution speed, using RelWithDebInfo should provide perfectly adequate game performance, while still retaining the debugging symbols.

**Comment by BenjamenMeyer** [comment]
We should generally ship with a `RelWithDebInfo` build config so folks can contribute usable trace stacks with information. If we do straight up `Release` builds then all the debug info is stripped away and it's useless to diagnostics.

As to local builds by devs - which is most useful really depends on the compiler you're using. If you're using the GNU Suite (gcc, g++) then it's just a matter of optimizations and whether the symbols are stripped. CLang and other OSS compilers are likely similar. However, the Microsoft Visual Studio C/C++ compiler adds some other features - Debug mode auto-initializes all pointers to 0xCCCCCCCCCCCCCCCC, zeros all memory allocations as well as disabling all optimizations, and does some other thread oriented stuff to help with multi-threaded development. Honestly for VS I recommend just using a Release Build and then disabling optimizations for debug so you have something usable that behaves like a delivered application. (Something I started doing even when I was still professionally using Visual Studios and delivering Windows applications.)

**Comment by stephengtuggy** [comment]
@BenjamenMeyer I just checked, and all our installer builds are set to build in `RelWithDebInfo` configuration. Shall I go ahead and close this ticket?

**Comment by BenjamenMeyer** [comment]
We did that a few years back. If there's nothing left to answer in this, then yes let's close it.
